### PR TITLE
chore: fix RUSTSEC-2024-0370 proc-macro-error is unmaintained

### PR DESCRIPTION
**Pull request by oleonardolima**
A candidate for fixing bitcoindevkit/bdk#1593. It builds on top of bitcoindevkit/bdk#1615 

### Description

In https://github.com/bitcoindevkit/bdk/issues/1593 it's mentioned that `proc-macro-error` is unmaintained for the
past few years, with no fix other than using proc-macro-error2 instead.

As on our scenario it's merely a transitive dependency of `clap`,
through `clap_derive` feature, which in latest releases doesn't depend on
`proc-macro-error` we can just bump it to latest.

It's valid to note that by bumping it, both examples that relies on clap
are no longer MSRV (1.63) compliant.

That said, this PR does:

- Standardize the example packages to have `example_` prefix.
- Exclude examples from running in main `Build & Test` CI job.
- Add new testing step to `Build & Test Example` CI job.
- Bumps the `clap` to `4.5.17`.

### Changelog notice

- Standardize the example packages to have `example_` prefix.
- Exclude examples from running in main `Build & Test` CI job.
- Add new testing step to `Build & Test Example` CI job.
- Bumps the `clap` to `4.5.17`.

#### All Submissions:

* [x] I've signed all my commits
* [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md)
* [x] I ran `cargo fmt` and `cargo clippy` before committing

 

#### Bugfixes:

* [ ] This pull request breaks the existing API
* [ ] I've added tests to reproduce the issue which are now passing
* [x] I'm linking the issue being fixed by this PR

### CODE REVIEW

**Comment by oleonardolima** [comment]
I also wonder if this could become even simpler by adopting [Examples Cargo Target](https://doc.rust-lang.org/cargo/reference/cargo-targets.html?highlight=examples#examples).

File: .github/workflows/cont_integration.yml [hunk 1]
```diff
@@ -50,9 +50,9 @@ jobs:
           cargo update -p tokio --precise "1.38.1"
           cargo update -p tokio-util --precise "0.7.11"
       - name: Build
-        run: cargo build ${{ matrix.features }}
+        run: cargo build --workspace --exclude 'example_*' ${{ matrix.features }}
```
Comment (ValuedMammal) [inline .github/workflows/cont_integration.yml:53]: We could try just doing `--all-features` 🤔

Comment (notmandatory) [reply .github/workflows/cont_integration.yml:53]: I think in the past we ran into errors that only happened when we built with `--no-default-features` so I'd like to keep it.

Comment (ValuedMammal) [reply .github/workflows/cont_integration.yml:53]: keep which one?

Comment (notmandatory) [reply .github/workflows/cont_integration.yml:53]: I mean keep testing with `--no-default-features` and `--all-features`. I don't mind only building the examples with the current `stable` rust though if that's the issue.

Comment (oleonardolima) [reply .github/workflows/cont_integration.yml:53]: I think it's better to keep testing with both.

**Comment by oleonardolima** [comment]
I think I also found a bug in how the `bdk_esplora` features were structured, it's somewhat depending on `blocking` feature being set up through the `wallet_esplora` example, and that's why it's failing in the CI when excluding the examples from the workspace.

**Comment by oleonardolima** [comment]
> I think I also found a bug in how the `bdk_esplora` features were structured, it's somewhat depending on `blocking` feature being set up through the `wallet_esplora` example, and that's why it's failing in the CI when excluding the examples from the workspace.

Also, it's something that we'd find out through bitcoindevkit/bdk#1944 if miniscript std/no-std weren't a blocker.

File: .github/workflows/cont_integration.yml [hunk 2]
```diff
@@ -172,3 +172,6 @@ jobs:
       - name: Build
         working-directory: example-crates/${{ matrix.example-dir }}
         run: cargo build
+      - name: Test
+        working-directory: example-crates/${{ matrix.example-dir }}
+        run: cargo test
```
Comment (ValuedMammal) [inline .github/workflows/cont_integration.yml]: Just noting I don't think there are any tests here to run.

Comment (oleonardolima) [reply .github/workflows/cont_integration.yml]: Oh, yes. Thank you!

**Comment by oleonardolima** [comment]
> I think I also found a bug in how the `bdk_esplora` features were structured, it's somewhat depending on `blocking` feature being set up through the `wallet_esplora` example, and that's why it's failing in the CI when excluding the examples from the workspace.

I was testing the fixes and got it in 8752555, but I'll move it to a separate PR to make it easier to review (#1615).

**Comment by ValuedMammal** [comment]
Rather than rename the example-crates can't we use the "exclude" key in the workspace Cargo.toml to exclude them from invocations of `cargo test --workspace`?

https://doc.rust-lang.org/cargo/reference/workspaces.html#the-members-and-exclude-fields

**Comment by oleonardolima** [comment]
> https://doc.rust-lang.org/cargo/reference/workspaces.html#the-members-and-exclude-fields

I thought about it, but AFAICT it'll exclude from the workspace for any `cargo` command, and not only for the `cargo test` one.

**Review by ValuedMammal — APPROVED** [review]
After considering the options I don't see any downside to using this solution.

**Comment by oleonardolima** [comment]
> After considering the options I don't see any downside to using this solution.

I rebased this one, as I updated the approach in bitcoindevkit/bdk#1615.

**Review by ValuedMammal — APPROVED** [review]
ACK d802d00a06035bec582eb5aadf0281f26bff6acd

**Pull request merged**